src/main/java/net/minecraftforge/common/capabilities/IAttachedCapabilityProvider.java

```diff
@@ -0,0 +1,107 @@
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.core.Direction;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.resources.ResourceLocation;
+
+/**
+ * An Attached Capability Provider is a simple capability provider that can be attached
+ * to game objects.<br>
+ * 
+ * 
+ * @param <C> The type of the capability.
+ * @param <O> The type of the provider being attached to.
+ */
+public interface IAttachedCapabilityProvider<C, O extends ICapabilityProvider>
+{
+
+    /**
+     * @return The type of capability this provider can return.
+     */
+    CapabilityType<C> getType();
+
+    /**
+     * @return The ID of this provider.
+     */
+    ResourceLocation getId();
+
+    /**
+     * Retrieves a capability from this provider.
+     * @param direction The side being queried. Will always be null if the underlying provider is not {@link ICapabilityProvider#isDirectionSensitive()}.
+     * @return The provided {@link Capability}, or {@link Capability#empty()} if not available for the <code>direction</code>
+     */
+    @NotNull Capability<C> getCapability(@Nullable Direction direction);
+
+    /**
+     * @see {@link ICapabilityProvider#invalidateCaps()} for documentation.
+     */
+    void invalidateCaps();
+
+    /**
+     * @see {@link ICapabilityProvider#reviveCaps()} for documentation.
+     */
+    void reviveCaps();
+
+    /**
+     * Serialize this provider to NBT, if applicable.<br>
+     * @return A {@link CompoundTag} representing the serialized form of this provider, or null, if not serializable.
+     */
+    default @Nullable CompoundTag serializeNBT()
+    {
+        return null;
+    }
+
+    /**
+     * Reads this provider from NBT.<br>
+     * This method will only be called if this provider was serialized via {@link #serializeNBT()}.
+     * @param tag The serialized tag from {@link #serializeNBT()}.
+     */
+    default void deserializeNBT(CompoundTag tag)
+    {
+    }
+
+    /**
+     * For {@link ItemStack}s to stack, it must be validated that they are equivalent.<br>
+     * This means validating that all attached capabilities are equivalent as well.<br>
+     * This method allows attached capabilities to define if they are equivalent or not.
+     * <p>
+     * Specifically, this method compares this provider to another instance of this provider
+     * on another object, or null (if it is not present on the other object).
+     * <p>
+     * The default implementation of this method always returns true, meaning this provider
+     * will be ignored during comparison.
+     * <p>
+     * If your capability must be equivalent for itemstacks to merge, you MUST override this method.
+     * 
+     * @param other Another instance of this provider from another object, or null, if it was not present.
+     * @return true if these providers are equivalent, otherwise false.
+     */
+    default boolean isEquivalentTo(@Nullable IAttachedCapabilityProvider<C, O> other)
```

Comment:
I think it would be reasonable for the default implementation to compare the results of serializing the caps, i.e.:
```java
if (other == null) return false;
return Object.equals(serializeCaps(new CompoundTag()), other.serializeCaps(new CompoundTag()))`
```

Then modders can override this if they have a more performant implementation that they wish to use.

Reply:
I think it might be better to force modders to define their own implementation because I think most would overlook overriding it purely for performance.

Reply:
I was thinking about providing that default, but then it is likely people utilize that heavily rather than write a proper compare function, which is why I omitted it.

Reply:
Imo making life easier for modders should be the priority of any API.

Reply:
I wouldn’t say that implementing a compare function isn’t easy. I also think performance is almost equally important in the case of this API.

Reply:
`areCapsCompatible` is not the performance issue at the moment, so it doesn't really make sense to ask modders to prematurely optimize it for now, imo.